### ISSUE TITLE
Multi User support

### ISSUE THREAD

**Issue by arielmoraes** [knadh/listmonk#543]
Does listmonk support multi users (each one with its own Dashboard)? 

If not, is it feasible to fork it and do the required changes?

**Comment by RKVodde** [comment]
Multi User Support is a much needed feature which boosts the usage of a wonderful platform like "listmonk" in a organizational setting.

**Comment by bryceprutsos** [comment]
I agree that multi-user support would be a great enhancement. However, like the developer said it's tricky because of which user has access to what. Just pointing out that if multi-user support is a thing LDAP or SSO (SAML, CAS, OpenID) should be supported.

**Comment by NicoHood** [comment]
For me it would be great to have a minimal multi user mode: Just have multiple logins, all with the same access. The reason why this is important, because you can withdraw permissions of the user (aka delete or block a user), when someone leaves the team. As a first step, this would help to get feedback and set the start for multiple logins.

**Comment by knadh** [comment]
+1 @NicoHood. There should be a simple native permission based multi-user system. Once that's in place, plugging in OAuth, LDAP should be straight forward.

**Comment by RKVodde** [comment]
Any progress on the multi user support.

**Comment by knadh** [comment]
@RKVodde nothing yet, unfortunately. I was looking at this again last week and came across this new lib: https://github.com/go-pkgz/auth

Haven't had a chance to look at it up close yet.

**Comment by RKVodde** [comment]
I am not into go-lang but this is what I found, might help. https://github.com/volatiletech/authboss

**Comment by c-nv-s** [comment]
> @RKVodde nothing yet, unfortunately. I was looking at this again last week and came across this new lib: https://github.com/go-pkgz/auth
> 
> Haven't had a chance to look at it up close yet.

I used this library before for a custom OAuth setup and it was fairly straightforward which is nice.
I actually switched to it after failing to get authboss working which was actually a painful experience.

**Comment by rpmcoding** [comment]
Sounds good!

**Comment by titansmc** [comment]
Hi,
I am interested on it but found no multi-user support. Is it something being worked on? Before we start evaluating alternatives.
Cheers.

**Comment by knadh** [comment]
It's on the to-do, but haven't started working on it yet. Unable to give a timeline, unfortunately.

**Comment by pwnoobz** [comment]
I'm also interested in this. I just want to stay subbed on the topic.

**Comment by jmaisonneuve01** [comment]
Hi I'm gonna start working on this function so if anybody got something already running message me plz!

**Comment by knadh** [comment]
@jmaisonneuve01 please share a plan/proposal/spec of what and how you're planning to implement.

**Comment by jmaisonneuve01** [comment]
> @jmaisonneuve01 please share a plan/proposal/spec of what and how you're planning to implement.

Still gathering and analyzing your code for the moment ill keep you updated

**Comment by samuk** [comment]
Any news on this one @jmaisonneuve01

**Comment by titansmc** [comment]
https://github.com/coreos/go-oidc

Any chance the OIDC could be used ?

**Comment by floridarangers** [comment]
For what it's worth until this feature is available. I use caddy and  forward_auth as a work around (nginx supports something like this as well). It's not perfect but it allows me to give multiple people access to the server without having this feature in place.

**Comment by knadh** [comment]
I've finally started working on this. Ref: https://github.com/knadh/listmonk/pull/1701#issuecomment-2032551568

**Comment by knadh** [comment]
Update:

![image](https://github.com/knadh/listmonk/assets/547147/9cea6f02-7eaf-4497-974e-35181dcbbdbf)

- User management on the UI and backend is done.
- WIP: Frontend authentication, user permissions, permissions per list.

**Comment by chrislbrown84** [comment]
Great news - thanks for sorting this @knadh

**Comment by bryceprutsos** [comment]
Great news @knadh I will be more than happy to test.

**Comment by knadh** [comment]
- User management, native password based login and OIDC login are done.
- WIP: Permissions

![image](https://github.com/knadh/listmonk/assets/547147/778945db-9a78-469c-99cd-17b57a334d5b)

**Comment by jflattery** [comment]
Looking good! Is the plan to implement this generically so that other OIDC providers (for example authentik) can be used?

**Comment by knadh** [comment]
The next release, v4.0.0 will have this. It should be ready in a month or slightly more than that.

**Comment by knadh** [comment]
![image](https://github.com/knadh/listmonk/assets/547147/934fd03b-1cc5-4f7d-8f96-ceefd3f1187a)

**Comment by carpepraedam** [comment]
@knadh 
Will this support Active Directory as an authentication source, or generic OIDC providers such as Keycloak.

Either of these would increase adoption from organizations operating within a corporate network.

**Comment by knadh** [comment]
OIDC support is already added. You can plug in any OIDC provider.

**Comment by shanrahul** [comment]
Will this support LDAP Authentication ?

**Comment by knadh** [comment]
There is no built-in support for LDAP that's planned. You can use something like KeyCloak or any proxy to achieve LDAP support.

**Comment by knadh** [comment]
WIP: Per-list read/write permissions attached to roles.

![image](https://github.com/knadh/listmonk/assets/547147/2e28a9ae-6e92-4e41-ad1f-c1c023067259)

**Comment by tudor-cel-stan** [comment]
Amazing work! Can't wait for the feature. When this is launched, maybe you could add the ability to connect SMTP configurations to lists, so that sending a campaign to a specific list would use that specific SMTP config. 

This would actually enable us to manage multiple clients/domains from a single ListMonk install. Thank you for building this tool!

**Comment by crebuh** [comment]
First of all great work @knadh 👏🏻👏🏻

to understand this feature correctly, it will not support different accounts for sending newsletters to different domains?

I guess this would be another feature or as @tudor-cel-stan mentioned, similar to having own smtp settings for different lists?

**Comment by knadh** [comment]
Thanks @crebuh

> it will not support different accounts for sending newsletters to different domains?

If you're referring to picking SMTP servers per campaign, that's not a part of the multi-user project.

**Comment by crebuh** [comment]
ok but multiple accounts for different companies/products?

**Comment by tudor-cel-stan** [comment]
@crebuh I was looking into quicker to implement ways to make Listmonk useable for multiple clients on the same domain. 

I think the quickest way to do that before starting to develop multitenancy (multiple organisations with different settings) which seems like a much larger project, is to be able to attach certain settings (like SMTP settings and sending address) to a mailing list.

For my client use cases, coupled with list permissions per user, the ability to change smtp and sending address settings on a per list basis would be ideal (I would be immediately switching my customers to Listmonk from Sendy to profit from your advanced Go templating).

**Comment by knadh** [comment]
Ah, there are no plans for building multi-tenancy into listmonk. That is outside the scope of the project. Also, one database (and app) that shares data for many orgs is not ideal, not in terms of maintenance, performance, or security when has not been built from the get-go to be that way.

If you want multi-tenancy for different clients, it's best to provision different instances of listmonk (it's light on CPU and RAM) with their own databases (can be one database server, but different databases init).

**Comment by crebuh** [comment]
@knadh Thanks for the clarification. i was thinking of a similar approach to add own instances of listmonk to all the different projects. normally this should be fine depending on the size of the project of course. and sharing the database is also the way to go.

**Comment by mustafa519** [comment]
That work looks amazing, but I am not sure if it helps to manage multiple projects/domains in the same listmonk instance. Does project/domain support count as a multi-tenancy implementation too?

I was just imagining project based subscriber separation to use the same instance for all my projects. Otherwise, having the same subscriber makes it harder to manage. Especially, there are many separated lists across projects.

I just mentioned the problem here that I am struggling right now #1959

Listmonk also can shine for the transactional mails like we can put the notification mails into the queue which listmonk is already supporting it.

However, I agree that it can double the DB entries since there will have to another relational table to build the relation between the domain and the subscriber, though. Still, I really need a decent solution to share the same listmonk instance across my projects.

**Comment by lmmendes** [comment]
@knadh Do you have any idea when this feature will be released?

**Comment by ssoronid** [comment]
> @crebuh ... For my client use cases, coupled with list permissions per user, the ability to change smtp and sending address settings on a per list basis would be ideal (I would be immediately switching my customers to Listmonk from Sendy to profit from your advanced Go templating).

For your use case, it would be beneficial to develop a system that manages subscribers effectively while maintaining data integrity and flexibility.

One approach would be to structure the database so that subscriber data is shared across the same database but associated with different lists through a "tenant" or "client" column. This would prevent conflicts when upserting subscribers who might be shared across multiple lists.

Here's how this could work:
- Tie each subscriber to a specific tenant or client using a unique identifier (e.g., a "tenant_id" column).
- When an upsert occurs (i.e., inserting or updating a subscriber), the system would use this tenant identifier to update the correct subscriber record for the appropriate list.
- This way, even if the same email address is subscribed to multiple lists, the system would update the appropriate data for each tenant or client, ensuring accurate and isolated subscriber management.

For SMTP configuration, you could store these settings directly in the tenant table. If a tenant has custom SMTP settings, they would be used; otherwise, the system could fall back to a default global configuration. This approach would offer the flexibility needed to tailor email sending per list while keeping the setup manageable.

Additionally, as the system evolves, you could consider implementing Row-Level Security (RLS) based on the `tenant_id` column. This would ensure that each tenant only accesses and modifies their own data, adding an extra layer of security to your database.

**Comment by knadh** [comment]
## Release candidate (RC3) for testing

This has taken forever, but listmonk v4.0.0 is almost ready for release. We've used it in our large production instance over the last two weeks without any obvious, noticeable flaws. Please test it out and share feedback. I am in the process of getting the new auth system professionally pen-tested.

This version involves major changes (while maintaining backwards compatibility), most importantly, the new authentication system with multi-user support.

-----------------------------

⚠️⚠️⚠️ **EXTREMELY IMPORTANT: Back up your Postgres database before upgrading, just to be safe!**

-----------------------------

## What changes?
The `admin_username` and `admin_password` fields from the config.toml are picked up and saved in the database under the new user management system by `--upgrade`. When you login after the upgrade, the dashboard will show a warning asking you to remove those fields from the TOML file. If you have listmonk API integrations in your programs, go to the new `Users` menu, create an API user, and use those credentials in the existing integrations before removing the old hardcoded credentials from TOML file.

## Upgrading to RC3 test version

### Docker
- Edit the Docker compose file and change the image name to `listmonk/listmonk:v4.0.0-rc-3`
- `docker compose pull` to pull the new image.
- Run database migrations with: `docker compose run --rm app ./listmonk --upgrade`
- Re-run with `docker compose up app db`

### Binary
- Replace your existing binary with the RC3 version: [listmonk.tar.gz](https://github.com/user-attachments/files/17117533/listmonk.tar.gz)
- Stop the running instance of listmonk.
- Run `./listmonk --upgrade`
- Start listmonk.

**Comment by rpmcoding** [comment]
@knadh are there any other new features in v4, or mainly just multi user support?  Not meant to sound ungrateful in any way, just interested.

**Comment by knadh** [comment]
@rpmcoding there are a number of small and big improvements and fixes: https://github.com/knadh/listmonk/compare/v3.0.0...multiuser

**Comment by paolospag** [comment]
Hello @knadh I'm curious to explore the 4.0.0 version with multi user support, then I tried to clone the repo from the branch 'multiuser' with this command:
`git clone -b multiuser https://github.com/knadh/listmonk.git`

but when I try to do:
`cd listmonk && make dist`

something goes wrong and I'm unable to compile binary. I got these errors:

```
CGO_ENABLED=0 go build -o listmonk -ldflags="-s -w -X 'main.buildString=v3.0.0 (#bf26ec8 2024-09-25T07:58:16+0000)' -X 'main.versionString=v3.0.0'" cmd/*.go
go: downloading github.com/zerodha/simplesessions/v3 v3.0.0
go: downloading github.com/coreos/go-oidc/v3 v3.9.0
go: downloading github.com/zerodha/simplesessions/stores/postgres/v3 v3.0.0
go: downloading golang.org/x/oauth2 v0.13.0
go: downloading github.com/go-jose/go-jose/v3 v3.0.1
# github.com/knadh/listmonk/models
models/models.go:828:32: undefined: PermListGetAll
models/models.go:833:32: undefined: PermListManageAll
models/models.go:846:19: undefined: PermListGet
models/models.go:850:19: undefined: PermListManage
make: *** [Makefile:42: listmonk] Error 1
```

Any idea?

**Comment by knadh** [comment]
@paolospag ouch, the new costs file wasn't committed. Please re-pull and retry.

**Comment by paolospag** [comment]
> @paolospag ouch, the new costs file wasn't committed. Please re-pull and retry.

I've tried right now and it works as expected.
Also I've run this without any problem:
`listmonk --upgrade --config /etc/listmonk/config.toml`

Thanks for your support, I'm excited to test this new version!!!

**Comment by EW1974** [comment]
Hi

An issue with redirections after login doing a fresh deployment. I also deployed v3.0.0 before in the same environment but different namespace without problems. We're looking forward for the multiuser version to avoid running separate instances of listmonk.

I've just used the helm chart to try v4.0.0-rc3 docker image in an isolated environment without internet access, so don't worry about messages about lacking access ;-) I don't know if the helm chart already reflects the needed changes for v.4.0.0 ...

It was deployed successfully and the listmonk pod is running. It's exposed using ingress, but when connecting via Browser and logging in the login page redirects absolutly to http://localhost:9000 instead of relative urls.

BTW -> Pod output log (main.go) still states v3.0.0 instead of v4.0.0. but I've rechecked that v4.0.0-rc3 image is being used 
```

2024/09/25 12:44:42 main.go:104: **v3.0.0** (#dc68be0 2024-09-20T16:46:54+0000)
2024/09/25 12:44:42 init.go:162: reading config: config.toml
2024/09/25 12:44:42 init.go:302: connecting to db: <NAME-REMOVED>:5432/listmonk
2024/09/25 12:44:42 init.go:658: media upload provider: filesystem
2024/09/25 12:44:42 init.go:1004: WARNING: Remove the admin_username and admin_password fields from the TOML configuration file. If you are using APIs, create and use new credentials. Users are now managed via the Admin -> Settings -> Users dashboard.
2024/09/25 12:44:42 init.go:581: loaded email (SMTP) messenger: [username@smtp.yoursite.com](mailto:username@smtp.yoursite.com)
⇨ http server started on 0.0.0.0:80
2024/09/25 12:59:42 updates.go:46: error checking for remote update: Get "https://update.listmonk.app/update.json": dial tcp: lookup update.listmonk.app on <REMOVED> : no such host
```

Containers /listmonk/config.toml
```

/listmonk # cat config.toml 
[app]
  admin_username = "admin"
  admin_password = "<REMOVED>"
  address = "0.0.0.0:80"

# Database.
  [db]
  host = "<REMOVED>"
  port = 5432
  user = "postgres"
  password = "<REMOVED>"
  database = "listmonk"
  ssl_mode = "disable"
  max_open = 25
  max_idle = 25
  max_lifetime = "300s"
```

Curl output and you can see that all the URLs are pointing to http://localhost:9000

```
$ curl -k https://<INGRESS-LOADBALANCER>/login/admin

<!DOCTYPE html>
<html>
<head>
        <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
        <title>listmonk - Mailing list</title>
        <meta name="description" content="" />
        <meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />


                <link rel="alternate" type="application/rss+xml" title="Mailing list archive - Mailing list"
                        href="http://localhost:9000/archive.xml" />


        <link href="/public/static/style.css?v=2fb65be82a" rel="stylesheet" type="text/css" />
        <link href="/public/custom.css?v=2fb65be82a" rel="stylesheet" type="text/css">
        <script src="/public/custom.js?v=2fb65be82a" async defer></script>


                <link rel="shortcut icon" href="/public/static/favicon.png?v=2fb65be82a" type="image/x-icon" />

</head>
<body>
    <div class="container wrap">
                <header class="header">
                        <div class="logo">
                                <a href="http://localhost:9000/subscription/form">

                                        <img src="/public/static/logo.svg?v=2fb65be82a" alt="listmonk" />

                                </a>
                        </div>
                </header>


<section class="center">
        <a href="admin" class="button">Login</a>

        <div class="home-options">

                        <a href="http://localhost:9000/subscription/form">Subscribe</a>


                    <a href="http://localhost:9000/archive">Mailing list archive</a>

        </div>
</section>


        </div>

        <footer class="container">
                Powered by <a target="_blank" rel="noreferrer" href="https://listmonk.app/">listmonk</a>
        </footer>
</body>
</html>


```

**Comment by knadh** [comment]
> It's exposed using ingress, but when connecting via Browser and logging in the login page redirects absolutly to http://localhost:9000/ instead of relative urls.

hm, unable to reproduce this. What do you see at `/admin/login`? Do you see the login form? Are you able to submit the form?

> BTW -> Pod output log (main.go) still states v3.0.0 instead of v4.0.0. but I've rechecked that v4.0.0-rc3 image is being used

Ah yes. This is because of the untagged RC compilation. This shouldn't affect functionality.

> $ curl -k https://<INGRESS-LOADBALANCER>/login/admin

It should be `/admin/login`.

**Comment by nbudin** [comment]
Thanks so much for this release @knadh!  I tried deploying it and connecting it to my organization's OIDC server; however, the scopes `profile` and `email` that are currently hard-coded into Listmonk RC3 don't exist in our server (we have them set up a bit differently).  Is there a plan to make the scopes configurable?  I'd be happy to work on a patch for this if it's helpful.

**Comment by knadh** [comment]
Thanks @nbudin. Aren't `email` and `profile` standard OIDC scopes (or at least [universal](https://auth0.com/docs/get-started/apis/scopes/openid-connect-scopes#standard-claims))? In your environment, what scopes communicate these pieces of information?

**Comment by nbudin** [comment]
> Thanks @nbudin. Aren't `email` and `profile` standard OIDC scopes (or at least [universal](https://auth0.com/docs/get-started/apis/scopes/openid-connect-scopes#standard-claims))? In your environment, what scopes communicate these pieces of information?

For us, it's `read_profile` because we've split out the scopes for read and write access.  That being said, I didn't realize this was part of the standard, so given that, I'd be okay with reconfiguring stuff on our end if you'd rather this not be configurable in Listmonk.

**Comment by knadh** [comment]
I checked a few popular apps that support OIDC and none seem to support remapping `email` and `profile` as they are standard fields. I don't think it would make sense to make it configurable @nbudin.

**Comment by sdjeff84** [comment]
Please help once upgrade it cannot start services.
Gives following error.
rror response from daemon: driver failed programming external connectivity on endpoint listmonk_demo_db (21a57ac1fd02f5d7d101c52b477f3d384543b1c16bf02dbbd0420ff600ed6012): Bind for 0.0.0.0:9432 failed: port is already allocated

**Comment by MaximilianKohler** [comment]
>Bind for 0.0.0.0:9432 failed: port is already allocated

You have to use a different port. You can also disable/remove the listmonk_demo section. You can search other issues in this project for more details.

**Comment by sdjeff84** [comment]
Dear @MaximilianKohler tried changing port, same error. Tried disable/remove demo section in docker compose.yml file no error but application status is restaring doesn't load. listmonk/listmonk:v4.0.0-rc-3   "docker-entrypoint.s…"   2 minutes ago   Restarting (1) please help.

**Comment by knadh** [comment]
Looks like you have multiple Postgres instances running, including the demo one. Do a docker compose down, remove the demo entries from the compose file, then do docker compose up.

**Comment by sdjeff84** [comment]
yes thats what i mentioned in my previous reply. did the same as you suggested this is the status listmonk/listmonk:v4.0.0-rc-3 "docker-entrypoint.s…" 2 minutes ago Restarting (1) .

**Comment by sdjeff84** [comment]
now its stucked here : 
root@mail:/opt/listmonk# docker compose up app db
WARN[0000] /opt/listmonk/docker-compose.yml: the attribute `version` is obsolete, it will be ignored, please remove it to avoid potential confusion
[+] Running 2/2
 ✔ Container listmonk_db   Running                                                                                   0.0s
 ✔ Container listmonk_app  Recreated                                                                                 0.4s
Attaching to listmonk_app, listmonk_db
listmonk_app  | Launching listmonk with user=[root] group=[root] PUID=[0] PGID=[0]
listmonk_app  | 2024/09/27 10:11:13 main.go:104: v3.0.0 (#dc68be0 2024-09-20T16:46:54+0000)
listmonk_app  | 2024/09/27 10:11:13 init.go:162: reading config: config.toml
listmonk_app  | 2024/09/27 10:11:13 init.go:302: connecting to db: listmonk_db:5432/listmonk
listmonk_app  | 2024/09/27 10:11:13 init.go:658: media upload provider: filesystem
listmonk_app  | 2024/09/27 10:11:13 init.go:1004: WARNING: Remove the admin_username and admin_password fields from the TOML configuration file. If you are using APIs, create and use new credentials. Users are now managed via the Admin -> Settings -> Users dashboard.
listmonk_app  | 2024/09/27 10:11:13 init.go:581: loaded email (SMTP) messenger: username@smtp.yoursite.com
listmonk_app  | ⇨ http server started on [::]:9000
Doesn't go beyond this have to kill the process.
And at the browser. after inserting user and password show site can't be reached.

**Comment by sdjeff84** [comment]
finally i figured it out. But when i login, i open with my ip address after inserting credentials its changing the address to localhost:9000/admin. and This site can’t be reached. please suggest

**Comment by TadeSF** [comment]
Really nice work @knadh, I love the improvements!
However, I just stumbled across something I find weird, but maybe it is intended behavior:
I connected an OIDC Issuer and got it working. I created an user with the email I was then going to use with the OIDC flow and made the user `Super Admin`. Then I went into incognito mode and logged in with OIDC and found myself – to my suprise – logged in as the default `@admin` user. 

Is this what should happen? Because in my understanding of how the OIDC Flow should work with listmonk is, that I create a user that can then log into their personal account with through OIDC. Otherwise (if the user does not exist yet) the login-flow should be rejected. Granting everyone the same access to the same user through OIDC doesn't seem like a good idea to me. Any Ideas?

**Comment by TadeSF** [comment]
> finally i figured it out. But when i login, i open with my ip address after inserting credentials its changing the address to localhost:9000/admin. and This site can’t be reached. please suggest

Had the same issue at first with my traefik setup.
**Solution**: Either ssh into the machine and use **port forwarding** to be able to reach the remote machines port 9000 on your computer (`ssh -L 9000:localhost:9000 user@host`) or – but I have not tried that yet and it could be problematic with cookies – replace the "localhost" in the link you were redirected to with the ip again.
After reaching the dashboard for the first time, you can easily change the `Root-URL` (Settings -> Settings -> General) to your ip to avoid any further redirection problems. Hope that helps!

**Comment by sdjeff84** [comment]
> @sdjeff84
> 
> > finally i figured it out. But when i login, i open with my ip address after inserting credentials its changing the address to localhost:9000/admin. and This site can’t be reached. please suggest
> 
> Had the same issue at first with my traefik setup. **Solution**: Either ssh into the machine and use **port forwarding** to be able to reach the remote machines port 9000 on your computer (`ssh -L 9000:localhost:9000 user@host`) or – but I have not tried that yet and it could be problematic with cookies – replace the "localhost" in the link you were redirected to with the ip again. After reaching the dashboard for the first time, you can easily change the `Root-URL` (Settings -> Settings -> General) to your ip to avoid any further redirection problems. Hope that helps!

yes tried replacing localhost with ip but after inserting the login credentials it again goes back to localhost. Need to know if through ssh if we can configure Root-URL.

**Comment by peterschmidler** [comment]
Hey thanks for the great work, really looking forward to installing it on my server. Any idea, when this feature will be merged to master?

**Comment by EW1974** [comment]
> > It's exposed using ingress, but when connecting via Browser and logging in the login page redirects absolutly to http://localhost:9000/ instead of relative urls.
> 
> hm, unable to reproduce this. What do you see at `/admin/login`? Do you see the login form? Are you able to submit the form?
> 
> > BTW -> Pod output log (main.go) still states v3.0.0 instead of v4.0.0. but I've rechecked that v4.0.0-rc3 image is being used
> 
> Ah yes. This is because of the untagged RC compilation. This shouldn't affect functionality.
> 
> > $ curl -k https:///login/admin
> 
> It should be `/admin/login`.

Yes this was just a typo of mine, but the problem persists, after login I'll get a redirect to https://localhost:9000

**Comment by smoebody** [comment]
> @sdjeff84
> 
> After reaching the dashboard for the first time, you can easily change the `Root-URL` (Settings -> Settings -> General) to your ip 

This should be a config variable, e.g. 
```
[app]
site_url = "http://localhost:9000"
```
and 
`LISTMONK_app__site_url=http://localhost:9000`

In this way the initial deployment can work without customising the setup.

**Comment by smoebody** [comment]
> Really nice work @knadh, I love the improvements! However, I just stumbled across something I find weird, but maybe it is intended behavior: I connected an OIDC Issuer and got it working. I created an user with the email I was then going to use with the OIDC flow and made the user `Super Admin`. Then I went into incognito mode and logged in with OIDC and found myself – to my suprise – logged in as the default `@admin` user.
> 
> Is this what should happen? Because in my understanding of how the OIDC Flow should work with listmonk is, that I create a user that can then log into their personal account with through OIDC. Otherwise (if the user does not exist yet) the login-flow should be rejected. Granting everyone the same access to the same user through OIDC doesn't seem like a good idea to me. Any Ideas?

I can confirm. After logging in i am always 'admin' user, regardless what 'preferred_username' or 'mail' attributes say.

**Comment by knadh** [comment]
Hi @smoebody. This was discovered fixed just yesterday (https://github.com/knadh/listmonk/commit/e7109daaf3a34d613b05133f23bdb2608bf6e10d). This was happening when OIDC claims sent a blank/invalid e-mail.

**Comment by smoebody** [comment]
@knadh thanks for your reply. I can confirm that login fails now that the email attribute is not available.

I guess you are not using the userinfo-endpoint to request the claimed attributes but instead trust upon that the OP sends the attributes in the id_token endpoint? Would it be much of an effort to implement the userinfo-enpoint or will I have to set up my OP accordingliy (which I desperately try for hours without luck [Shibboleth-IdP by the way]).

**Comment by knadh** [comment]
@smoebody I've added a patch (https://github.com/knadh/listmonk/commit/25cdb7b18eb41821d76e99ae98929e2b5d714b96) that pulls e-mail from the userinfo endpoint if the token claim doesn't have it (I wasn't aware that this was a convention).

Would you be able to test the [oidc-userinfo](https://github.com/knadh/listmonk/tree/oidc-userinfo) branch in your environment?

**Comment by knadh** [comment]
Got a professional pentest done on the RC, which took some time.

Just added the last remaining bit of UX, prompting for new user setup on fresh installs. 

This release has taken forever, but I'm finally finishing up the docs and pushing the release tomorrow.

![image](https://github.com/user-attachments/assets/5e3c4be7-9733-4e46-91e6-7681be9d9caa)

**Comment by knadh** [comment]
[v4.0.0](https://github.com/knadh/listmonk/releases/tag/v4.0.0) is out. Finally, whew. Thank you for your patience!

**Comment by BloodyIron** [comment]
Not sure if this is the place for this request but can we get User Auth expanded to support LDAP/AD and SAML please? OIDC as the only central-auth protocol being supported is very limiting in terms of which environments this can work in.

**Comment by alfredosola** [comment]
> Instead of adding just more auth methods, I would like to see the focus on improving the existing one (for example by adding optional OIDC features like backchannel-logout, etc). Users with LDAP or SAML sources can still use software like Keycloak, or SATOSA to proxy between their auth backend and OIDC. As the saying is: "Do one thing and do it well"

Further and supporting your point, there is the possibility to setup an authentication proxy. In the past, I have used [dex](https://github.com/dexidp/dex) and that works just fine: it is light and easy to setup. Do OIDC well, and if LDAP/SAML1/whatever is needed, just use a proxy.

**Comment by knadh** [comment]
+1 to @ErikMichelson. It doesn't make sense to add so many auth protocols (especially in a newsletter manager). If one really needs SAML, LDAP, or any other auth layer, then using an auth proxy is the right approach.

> for example by adding optional OIDC features like backchannel-logout, etc

Waiting for https://github.com/coreos/go-oidc/pull/251 to be merged.

**Comment by BloodyIron** [comment]
> Hot take, but both LDAP and SAML can be called "authentication methods from the last decade" (although I know how widespread they are used). Implementing either of them correctly can be a larger task, and implementing them partly would either bring security weaknesses or incompatibilities with some IdPs and in turn more support questions about setting them up correctly.
> 
> Instead of adding just more auth methods, I would like to see the focus on improving the existing one (for example by adding optional OIDC features like backchannel-logout, etc). Users with LDAP or SAML sources can still use software like Keycloak, or SATOSA to proxy between their auth backend and OIDC. As the saying is: "Do one thing and do it well"

Implementing LDAP/AD interfacing would not add security vulnerabilities lol that's absurd. LDAP and AD interfacing can be done securely via TLS with Bind user authentication. You're creating FUD here bud, Fear Uncertainty and Doubt from a baseless position. Also... SAML is from the last decade, and is insecure? Yeah this isn't accurate in the slightest.

Improving existing forms of authentication is something I can agree with, but blocking the completely legitimate asks for other authentication methods to be added based on inaccurate information is just uncalled for and out of line.

**Comment by ErikMichelson** [comment]
@BloodyIron Please keep the discussion on a professional level without accusing other people of acting in bad interest.

I'm honestly tired of responding to messages like this, but we'd need to clarify the facts.

I neither wrote “LDAP/AD is insecure” nor “SAML is insecure”.
Implementing these features correctly *can* be a challenge and if there are problems with the implementation this has likely a large impact on the security of the application as a whole — like for every other auth provider (including OIDC) too. Therefore, focussing on one authentication method is a reasonable suggestion. I'm not arguing that OIDC is more widespread used than LDAP, but first of all this is already implemented in Listmonk, and second it's used by major players like Google, Apple or Auth0.

LDAP/AD can be used via TLS, that's correct. Still, based on the implementation (either directly or using a 3rd-party library) there are risks like e.g., injection of user input when using more complex queries for filtering the access based on more criteria. See [CVE 2021-41232](https://nvd.nist.gov/vuln/detail/CVE-2021-41232) for one specific example or [CWE-90](https://cwe.mitre.org/data/definitions/90.html) for the general problematic.

SAML is based upon XML, which has its own fair share of problems due to a complex parsing process. See the CWEs [91](https://cwe.mitre.org/data/definitions/91.html) or [112](https://cwe.mitre.org/data/definitions/112.html) with several CVEs matching these, as an example. The verification of signatures in SAML responses is another common source for weaknesses in SAML. See [CVE-2024-45409](https://nvd.nist.gov/vuln/detail/CVE-2024-45409) or [CVE-2024-8698](https://nvd.nist.gov/vuln/detail/CVE-2024-8698) as examples. That a properly secure SAML implementation is not a no-brainer, show rather recent vulnerabilities, like with [GitHub Enterprise Server](https://cybersecuritynews.com/saml-bypass-authentication-on-github-enterprise-servers/), [GitLab's SAML support](https://cert.europa.eu/publications/security-advisories/2024-101/pdf) or [Germany's national auth system](https://www.heise.de/en/news/Security-flaw-in-municipal-website-causes-data-leak-in-BundID-9770296.html).

Again, I'm not saying that OIDC is problem-free. There are critical issues too with some implementations ([CVE-2024-10318](https://nvd.nist.gov/vuln/detail/CVE-2024-10318), [CVE-2021-44878](https://nvd.nist.gov/vuln/detail/CVE-2021-44878), just to showcase some examples). But keeping an eye on possible problems for one authentication method is better doable than keeping an eye on possible problems for more authentication methods.

With this, I'm not open for further discussion, especially in this already closed issue of a great open-source project. Instead, let's thank @knadh for maintaining this project and adding features like the multi-user support.

To not further disturb people who have subscribed notifications for this issue, I'd propose to lock it and instead create new ones for further suggestions - but that's ultimately the maintainers' decision.

**Comment by knadh** [comment]
Ah, I'd forgotten that this thread is almost ~5 years old. Let's open a new issue if a discussion is warranted. That said, as I mentioned earlier, I don't think this it's viable to add and maintain more auth implementations in listmonk. Thankfully, auth proxies solve that issue (albeit with some extra effort). Thanks everyone.
